fix(wcag): notification, overflowMenu & search accessibility enhancements

jaideep08 opened the pull request:
## Summary

Brief summary describing the changes here.

Made enhancements to following components to support voiceOver or screenReader accessibility.
- Notification
- Overflow menu
- Search input

## ADO Story or GitHub Issue Link

[USER STORY 1904498](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1904498)
[USER STORY 1904500](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1904500)

## Checklist

- [x] Used Conventional Commit messages as outlined in the contributing guide.
  - [ ] Noted breaking changes (if any).
- [x] Documented/updated all props, events, slots, parts, etc with JSDoc.
  - [x] Ran the `analyze` command to update Storybook docs.
- [x] Added/updated Stories with controls to cover all variants.
- [x] Ran `test` locally to address any failures.
- [ ] Added/updated the Figma link for the Story's Design tab.
- [ ] Added any new component exports to the src/index.ts file

## Testing Instructions

Use the system's voiceOver or screenReader for testing.
**Note**: _Please refer to the description provided in the tasks linked to the above mentioned stories._

## Screenshots

**Search input**
<img width="1728" alt="Screenshot 2024-09-12 at 11 05 38 AM" src="https://github.com/user-attachments/assets/ae038906-79e7-46b5-957a-1f83ddc42e92">

**Notification**
<img width="1728" alt="Screenshot 2024-09-12 at 11 12 51 AM" src="https://github.com/user-attachments/assets/404293f4-3034-4945-8e8c-470d7c1a03f4">
<img width="1728" alt="Screenshot 2024-09-12 at 11 13 23 AM" src="https://github.com/user-attachments/assets/f47177f1-92b6-43fe-a39b-916b87724789">
<img width="1727" alt="Screenshot 2024-09-12 at 11 14 05 AM" src="https://github.com/user-attachments/assets/7a1af5a3-bb7f-44f9-bd60-41fa99830c2f">
<img width="1726" alt="Screenshot 2024-09-12 at 11 14 38 AM" src="https://github.com/user-attachments/assets/5727a94d-082d-4990-a963-e1e21b30b3a7">
<img width="1728" alt="Screenshot 2024-09-13 at 12 50 25 PM" src="https://github.com/user-attachments/assets/662ca4a7-8327-483a-a801-8b61351db0a2">


**Overflow menu**

https://github.com/user-attachments/assets/467c6d47-e74e-4266-99e3-80c770a02087

brian-patrick-3 commented on src/components/reusable/tag/tag.ts:
```diff
@@ -56,6 +56,12 @@ export class Tag extends LitElement {
+   * Determines if the tag is used in the notification component.
+   */
+  @property({ type: Boolean })
+  isNotification = false;
```
We should use more generic property names like `description` instead, as components are isolated and do not relate to each other.

brian-patrick-3 reviewed (changes requested):
Had a few comments

brian-patrick-3 commented on src/components/reusable/notification/notification.ts:
```diff
@@ -148,15 +170,16 @@ export class Notification extends LitElement {
                 slot="icon"
                 fill=${notificationIconFillColor[this.tagStatus]}
                 .icon=${notificationIcon[this.tagStatus]}
+                iconDescription=${`${this.textStrings[this.tagStatus]} icon`}
```
Ignore this, just saw that is the other PR.

brian-patrick-3 commented on src/components/reusable/notification/notification.ts:
```diff
@@ -115,9 +135,11 @@ export class Notification extends LitElement {
             rel="noopener"
             @on-card-click=${(e: any) => this._handleCardClick(e)}
             hideBorder
+            cardRole="alertdialog"
+            cardDescription=${this.assistiveNotificationTypeText}
```
Ignore this, just saw that is the other PR.

jaideep08 commented on src/components/reusable/notification/notification.ts:
```diff
@@ -115,9 +135,11 @@ export class Notification extends LitElement {
             >${this.renderInnerUI()}</kd-card
           >`
-        : html`<kd-card type=${this.type} class="${classMap(cardBgClasses)}"
+        : html`<kd-card type=${this.type} cardRole="alertdialog" cardDescription=${this.assistiveNotificationTypeText} class="${classMap(cardBgClasses)}"
```
there is an accessibility violation that says all notifications need to have ARIA role as `alertdialog`. Hence added.

jaideep08 commented on src/components/reusable/notification/notification.ts:
```diff
@@ -59,6 +59,26 @@ export class Notification extends LitElement {
+
+  /** Assistive text for notification type (Required to support accessibility). */
+  @property({ type: String })
+  assistiveNotificationTypeText = '';
```
there's an accessibility requirement which says, colour is used as additional information to show the type of notification. The type of notification is not read out. Hence added.

jaideep08 commented on src/components/reusable/notification/notification.ts:
```diff
@@ -59,6 +59,18 @@ export class Notification extends LitElement {
+
+  /** Assistive text for timestamp (Required to support accessibility). */
+  @property({ type: String })
+  assistiveTimestampText = '';
```
kept this with respect to the following accessibility requirement,
_the duration text should be separately declared as "Duration", just reading the time, someone using screenreader might miss the context_

brian-patrick-3 replied:
Maybe we should default this to 'Duration' then

jaideep08 commented:
@brian-patrick-3, i've addressed some review comments, added a few comments and added support for overflow menu keyboard navigation. Please have a look.

kumarivijeta commented on src/components/reusable/notification/notification.ts:
```diff
@@ -170,14 +191,16 @@ export class Notification extends LitElement {
                   fill="#3D3C3C"
                   .icon=${closeIcon}
+                  role="img"
+                  aria-label="Close icon"
```
No hard-coded strings, must be customizable for localization purposes. Instead of "Close icon"can we  just have "Close" text?

kumarivijeta replied:
Can also add a title attribute here ?

kumarivijeta commented on src/components/reusable/notification/notification.ts:
```diff
@@ -198,11 +221,12 @@ export class Notification extends LitElement {
                 label=${this.textStrings[this.tagStatus]}
                 tagColor=${this._tagColor[this.tagStatus]}
                 shade="dark"
+                aria-label="Status"
```
No hard-coded strings, must be customizable for localization purposes.

kumarivijeta reviewed (approved):
LGTM

srpriyankashetty commented:
Using NVDA - 
Notification - The duration is reading just the time not saying it as Duration
Notification status - just reading as Success - not saying as Status

jaideep08 commented:
> Using NVDA - Notification - The duration is reading just the time not saying it as Duration Notification status - just reading as Success - not saying as Status

The aria-label has been added on both the tag and the timeStamp element and the mac voiceOver is reading them out just fine. The issue seems to be with other screen reader tools, not sure why.
<img width="1728" alt="Screenshot 2024-09-12 at 4 00 03 PM" src="https://github.com/user-attachments/assets/6c01535a-ea7e-41bb-9cf9-1afd54bad827">
<img width="1728" alt="Screenshot 2024-09-12 at 4 00 34 PM" src="https://github.com/user-attachments/assets/3972444a-c27e-47c1-b93f-4a4f308c7031">

srpriyankashetty commented:
Using NVDA able to see only the assistiveNotificationTypeText. The notificationRole is not read by the reader i.e Alert, Status or Log. In MAC its working fine.

![image](https://github.com/user-attachments/assets/b0a30c1c-95a8-4720-98d2-6f1324ccdcbf)

brian-patrick-3 commented:
> Using NVDA able to see only the assistiveNotificationTypeText. The notificationRole is not read by the reader i.e Alert, Status or Log. In MAC its working fine.

I don't think we would necessarily expect the "role" to be read out loud, it is more to give the screen reader semantic context, so this probably fine as-is.
